corpus/convert.py

```diff
@@ -93,6 +124,61 @@ def run(self):
         c = corpus.Corpus()
         c.load(corpus_path)
 
-        with open(self.txt_path.get_path(), "wt") as out:
-            for segment in c.segments():
+        if self.segment_file:
+            with open(tk.uncached_path(self.segment_file), "rt") as f:
+                segments_whitelist = set([l.strip() for l in f.readlines() if len(l.strip()) > 0])
+        else:
+            segments_whitelist = None
+
+        if self.gzip:
+            out = gzip.open(self.out_txt.get_path(), "wt")
+        else:
+            out = open(self.out_txt.get_path(), "wt")
+
+        for segment in c.segments():
+            if not segments_whitelist or segment.fullname in segments_whitelist:
                 out.write(segment.orth + "\n")
+
+        out.close()
+
+
+class CorpusReplaceOrthFromTxtJob(Job):
+    """
+    Merge raw text back into a bliss corpus
+    """
+
+    def __init__(self, corpus, text_file, segment_file=None):
+        """
+
+        :param Path corpus: Bliss corpus
+        :param Path text_file: a raw or gzipped text file
+        :param Path|None: only replace the segments as specified in the segment file
+        """
+        self.corpus_path = corpus
+        self.text_file = text_file
+        self.segment_file = segment_file
+
+        self.out_corpus = self.output_path("corpus.xml.gz")
+
+    def tasks(self):
+        yield Task("run", mini_task=True)
+
+    def run(self):
+        c = corpus.Corpus()
+        c.load(tk.uncached_path(self.corpus_path))
+
+        if self.segment_file:
+            with open(tk.uncached_path(self.segment_file), "rt") as f:
+                segments_whitelist = set([l.strip() for l in f.readlines() if len(l.strip()) > 0])
```

Comment:
you are still creating an intermediate list, you can create the set from an iterator, no need for the list